crypto/x509/x509spki.c

```diff
@@ -131,3 +132,45 @@ char *NETSCAPE_SPKI_b64_encode(NETSCAPE_SPKI *spki) {
   OPENSSL_free(der_spki);
   return b64_str;
 }
+
+int NETSCAPE_SPKI_print(BIO *out, NETSCAPE_SPKI *spki) {
+  if (out == NULL || spki == NULL) {
+    OPENSSL_PUT_ERROR(X509, ERR_R_PASSED_NULL_PARAMETER);
+    return 0;
+  }
+  BIO_printf(out, "Netscape SPKI:\n");
+
+  // Print out public key algorithm and contents.
+  ASN1_OBJECT *spkioid;
+  X509_PUBKEY_get0_param(&spkioid, NULL, NULL, NULL, spki->spkac->pubkey);
+  BIO_printf(out, "  Public Key Algorithm: %s\n",
+             (OBJ_obj2nid(spkioid) == NID_undef)
+                 ? "UNKNOWN"
+                 : OBJ_nid2ln(OBJ_obj2nid(spkioid)));
+  EVP_PKEY *pkey = X509_PUBKEY_get0(spki->spkac->pubkey);
+  if (pkey == NULL) {
+    BIO_printf(out, "  Unable to load public key\n");
+  } else {
+    EVP_PKEY_print_public(out, pkey, 4, NULL);
+  }
+
+  ASN1_IA5STRING *chal = spki->spkac->challenge;
+  if (chal->length != 0) {
+    BIO_printf(out, "  Challenge String: %.*s\n", chal->length, chal->data);
+  }
+
+  // Print out signature algorithm and contents.
+  BIO_printf(out, "  Signature Algorithm: %s",
+             (OBJ_obj2nid(spki->sig_algor->algorithm) == NID_undef)
+                 ? "UNKNOWN"
+                 : OBJ_nid2ln(OBJ_obj2nid(spki->sig_algor->algorithm)));
+  for (int i = 0; i < spki->signature->length; i++) {
+    if ((i % 18) == 0) {
+      BIO_write(out, "\n      ", 7);
```

Comment:
nit: suggest pulling the string out into a `const` and calling `strlen()` on it in the `BIO_write` invocation

or perhaps consider using `BIO_printf`

Reply:
Great point, I automatically assumed 7 was an indentation.. Changed to use `BIO_printf` instead